File: Brainfuck/brainfuck.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stack>
#include <unordered_map>

int main() {
	std::string program;
	std::cin >> program;
	
	std::vector<unsigned char> tape = {0};
	size_t cell_index = 0;
	
	std::string user_input;
	std::unordered_map<size_t, size_t> loop_table;
	
	char instruction;
	size_t loop_beginning_index;
	std::stack<size_t> loop_stack;
	for(size_t ip = 0; ip < program.size(); ip++) {
		instruction = program[ip];
		if(instruction == '[') {
			loop_stack.push(ip);
		} else if(instruction == ']') {
			loop_beginning_index = loop_stack.top();
			loop_stack.pop();
			loop_table[loop_beginning_index] = ip;
			loop_table[ip] = loop_beginning_index;
		}
	}
	
	size_t ip = 0;
	while(ip < program.size()) {
		instruction = program[ip];
		
		switch(instruction) {
			case '+':
				tape[cell_index] += 1;
				break;
			case '-':
				tape[cell_index] -= 1;
				break;
			case '<':
				cell_index--;
				break;
			case '>':
				cell_index++;
				if(cell_index == tape.size()) {
					tape.push_back(0);
				}
				break;
			case '.':
				std::cout << tape[cell_index];
				break;
			case ',':
				if(user_input.empty()) {
					std::cin >> user_input;
					user_input.push_back('\n');
				}
				tape[cell_index] = user_input[0];
				user_input.erase(0, 1);
				break;
			case '[':
				if(!tape[cell_index]) {
					ip = loop_table[ip];
				}
				break;
			case ']':
				if(tape[cell_index]) {
					ip = loop_table[ip];
				}
				break;
		}
		
		ip++;
	}
}

File: Brainfuck/brainfuck.py
program = input()

tape = [0]
cell_index = 0

user_input = []
loop_table = {}

loop_stack = []
for ip, instruction in enumerate(program):
	if instruction == "[":
		loop_stack.append(ip)
	elif instruction == "]":
		loop_beginning_index = loop_stack.pop()
		loop_table[loop_beginning_index] = ip
		loop_table[ip] = loop_beginning_index

ip = 0
while ip < len(program):
	instruction = program[ip]
	
	if instruction == "+":
		tape[cell_index] += 1
		if tape[cell_index] == 256:
			tape[cell_index] = 0
	elif instruction == "-":
		tape[cell_index] -= 1
		if tape[cell_index] == -1:
			tape[cell_index] = 255
	elif instruction == "<":
		cell_index -= 1
	elif instruction == ">":
		cell_index += 1
		if cell_index == len(tape):
			tape.append(0)
	elif instruction == ".":
		print(chr(tape[cell_index]), end="")
	elif instruction == ",":
		if user_input == []:
			user_input = list(input() + "\n")
		tape[cell_index] = ord(user_input.pop(0))
	elif instruction == "[":
		if not tape[cell_index]:
			ip = loop_table[ip]
	elif instruction == "]":
		if tape[cell_index]:
			ip = loop_table[ip]
	
	ip += 1